packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3396 @@
+    properties.add(DiagnosticsProperty<_MenuNode?>('parent', isRoot ? null : parent, defaultValue: null));
+  }
+}
+
+class _RootMenuNode extends _MenuNode {
+  _RootMenuNode(this.controller) : menuScopeNode = FocusScopeNode();
+
+  @override
+  void dispose() {
+    menuScopeNode.dispose();
+    super.dispose();
+  }
+
+  @override
+  MenuController controller;
+
+  // A list of descendant menus that are currently open.
+  final Set<_MenuNode> openMenus = <_MenuNode>{};
+
+  @override
+  bool get isRoot => true;
+
+  @override
+  _RootMenuNode get root => this;
+
+  @override
+  _MenuNode get parent => throw UnimplementedError('Tried to get the parent of the root node');
+
+  @override
+  Axis get orientation => Axis.horizontal;
+
+  @override
+  Offset? get globalMenuPosition => null;
+
+  @override
+  FocusScopeNode menuScopeNode;
+
+  @override
+  void focusButton() {}
+}
+
+class _ChildMenuNode extends _MenuNode {
+  _ChildMenuNode({
+    required this.parent,
+    required MenuController controller,
+    required List<Widget> widgetChildren,
+    required GlobalKey buttonKey,
+    FocusNode? buttonFocusNode,
+    this.onOpen,
+    this.onClose,
+    Offset alignmentOffset = Offset.zero,
+    Offset? globalMenuPosition,
+    Axis orientation = Axis.vertical,
+    ButtonStyle? buttonStyle,
+    MenuStyle? menuStyle,
+    Clip menuClipBehavior = Clip.none,
+    MaterialStatesController? statesController,
+  })  : _buttonFocusNode = buttonFocusNode,
+        _buttonKey = buttonKey,
+        _controller = controller,
+        _statesController = statesController,
+        _widgetChildren = widgetChildren,
+        _orientation = orientation,
+        _alignmentOffset = alignmentOffset,
+        _globalMenuPosition = globalMenuPosition,
+        _buttonStyle = buttonStyle,
+        _menuStyle = menuStyle,
+        _menuClipBehavior = menuClipBehavior,
+        menuScopeNode = FocusScopeNode() {
+    assert((){
+      menuScopeNode.debugLabel = 'Menu Scope';
+      return true;
+    }());
+    parent.addChild(this);
+  }
+
+  @override
+  MenuController get controller => _controller;
+  MenuController _controller;
+  set controller(MenuController value) {
+    if (_controller != value) {
+      _controller = value;
+      _notifyNextFrame();
+    }
+  }
+
+  MaterialStatesController? get statesController => _statesController;
+  MaterialStatesController? _statesController;
+  set statesController(MaterialStatesController? value) {
+    if (_statesController != value) {
+      _statesController = value;
+      _notifyNextFrame();
+    }
+  }
+
+  @override
+  bool get isRoot => false;
+
+  @override
+  _MenuNode parent;
+
+  OverlayEntry? overlayEntry;
+  bool isOpen = false;
+  VoidCallback? onOpen;
+  VoidCallback? onClose;
+
+  Offset get alignmentOffset => _alignmentOffset;
+  Offset _alignmentOffset;
+  set alignmentOffset(Offset value) {
+    if (_alignmentOffset != value) {
+      _alignmentOffset = value;
+      _notifyNextFrame();
```

Comment:
Does that mean that any change to the menu will be delayed by one frame? Why is that necessary? If I change the alignment in a button handler, it should become active the very next frame? Not two frames later?

Reply:
Yes, it will be delayed by a frame.  The menu bar or the widget that you've attached the menu to is in the main widget tree.  The menu is in the overlay, so marking it dirty needs to happen between frames.

Reply:
What if I am already between frames though (e.g. in an onpressed handler of a button) when reconfiguring the Node? In that scenario, updates will now be delayed by two frames, I think. Timeline:

(we start being between to frames)
* Node is reconfigured, _notifyNextFrame schedules a postframe callback (P) and requests a frame (F1)
* Scheduled frame (F1) happens
* At the end of (F1), (P) executes and listeners of the node are informed
* Listeners dirty the tree to adapt to the updated node, which schedules another frame F2
* F2 is produced to show the updated node configuration on screen.

THe update is now two frames late. We could have applied it in F1, though.

I think you just need to check in _notifyNextFrame whether we are between frames and if we are notifyLIsteners directly, if we are inside a frame, do the postframecallback thing.